diff --git a/src/index.js b/src/index.js
index 56f267a931ea1ff54ccd353d577afcc37f7bc9db..1d529cae44ebb7f50a9e6a18ced7e3ba1fde8fe6 100644
--- a/src/index.js
+++ b/src/index.js
@@ -39,10 +39,6 @@ class Storage {
     this.options.up = options
     await this.preCreate(directory, this.options)
 
-    if (!this.storage) {
-      if (fs && fs.mkdirSync) fs.mkdirSync(directory, { recursive: true })
-    }
-
     const db = new Level(directory, options)
     await db.open()
 
@@ -52,7 +48,7 @@ class Storage {
   async destroy (store) {
     if (!this.storage || !this.storage.destroy) return
 
-    await this.storage.destory(store._db.location)
+    await this.storage.destroy(store._db.location)
   }
 
   async preCreate (directory, options) {} // to be overridden