--- /home/adroit/.elm/0.19.1/packages/elm-explorations/webgl/1.1.3/src/Elm/Kernel/WebGL.js	2024-06-04 20:49:20.496093720 -0500
+++ /home/adroit/.elm/0.19.1/packages/elm-explorations/webgl/1.1.3/src/Elm/Kernel/WebGL_patched.js	2024-06-04 20:49:51.076047240 -0500
@@ -787,6 +787,7 @@
   }, model.__options);
 
   var canvas = __VirtualDom_doc.createElement('canvas');
+  canvas.created_by_elm = true;
   var gl = canvas.getContext && (
     canvas.getContext('webgl', options.contextAttributes) ||
     canvas.getContext('experimental-webgl', options.contextAttributes)
@@ -834,6 +835,7 @@
 
   } else {
     canvas = __VirtualDom_doc.createElement('div');
+    canvas.created_by_elm = true;
     canvas.innerHTML = '<a href="https://get.webgl.org/">Enable WebGL</a> to see this content!';
   }
 
